README
SBLIM Java CIM Client
=====================

(C) Copyright IBM Corp. 2005, 2009

 THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE 
 ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE 
 CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.
 
 You can obtain a current copy of the Eclipse Public License from
 http://www.opensource.org/licenses/eclipse-1.0.php

Author:             Viktor Mihajlovski <mihajlov@de.ibm.com>
Contributors:
                    Thorsten Schaefer  <thschaef@de.ibm.com>
                    Wolfgang Taphorn   <taphorn@de.ibm.com>

Last Change Date: 06/22/2009


About this Package
------------------
The purpose of this package is to provide a CIM Client Class Library
for Java applications. It complies to the DMTF standard CIM Operations over
HTTP and intends to be compatible with JCP JSR48 once it becomes available.
To learn more about DMTF visit http://www.dmtf.org.
More infos about the Java Community Process and JSR48 can be found at
http://www.jcp.org and http://www.jcp.org/en/jsr/detail?id=48.


Requirements
------------
In order to build, install and operate this package you will need a 
Java SDK or Run Time Environment of version 1.4 or higher, depending on
whether you build binaries from sources or use the pre-built binaries (JAR
file).

Sun Java SDK/JRE:
For users of the Sun Java packages, the following package is required to be 
adapted into the CLASSPATH environment variable:
 - xerces    http://xml.apache.org/xerces2-j

IBM Java SDK/JRE:
Because the IBM Java packages includes already the xerces package, no further
requirements need to be resolved.

As a requirement for building the SBLIM CIM Client, Apache Ant 
(http://ant.apache.org) needs to be setup properly. We would like to recommend
a Apache Ant version 1.6.5 or higher. (Lower versions might also work properly
here)


Building and Installing the CIM Client
--------------------------------------
The build process is based on Apache Ant, so it is very easy to build the 
SBLIM CIM Client.

Running Sun Java SDK:
If you have a xerces parser included into your systems CLASSPATH environment 
variable, no further dependencies need to be resolved.
If you don't have a xerces parser included, please create a directory named 
'lib' in the 'basedir' directory of the SBLIM CIM Client. Please copy the 
'xerces' Java Archives (JAR files) into this 'lib' directory.

An Ant build.xml file is in place which supports the following targets:
  
  help          : This target can be used to display a description of the several 
                  targets.
  clean         : This target can be used to clean up the project from previous
                  builds.
  build         : This target can be used to build the Java classes of the SBLIM
                  CIM Client project.
  build-release : This target can be used to build the Java classes of the SBLIM
                  CIM Client project and also the JavaDoc.
  java-doc      : This target can be used to compose the JavaDoc for all SBLIM
                  CIM Client classes.
  package       : This target can be used to build a Java Archive (JAR file) 
                  which includes a Manifest and all Java classes of the SBLIM CIM
                  Client.
  release       : This target can be used to build zip archives and tar balls 
                  containing source, binary and documentation of the SBLIM CIM 
                  Client.
  cvs-update    : This target can be used to update the current SBLIM CIM Client
                  to the latest version from CVS.

The execution of the build process is very easy. Once, Apache Ant is setup 
properly on your system, just execute 'ant <target>' in the 'basedir' directory
of the SBLIM CIM Client. The output will be written into a new created
directory named 'cim-client'. That's all.


Using the SBLIM CIM Client
--------------------------
The SBLIM CIM Client is a Java implementation that can be used in all 
applications which deal with CIMOMs.
It is packaged into a Java Archive (JAR file) that can be consumed by other
Java applications as needed.

sblimCIMClient.jar:	This jar contains the core client with the SBLIM API

Additionally, the SBLIM CIM Client is also packaged as RPM package which can be
used in Linux Distributions to be installed. The RPM is based on the 
infrastructure that is defined by the JPackage project.


NEWS
Changes in 1.3.9.3
==================
3160431 Need to support property file on AIX.

Changes in 1.3.9.2
==================
3078216 Fix for a null pointer exception in 1.3.9.1

Changes in 1.3.9.1
==================
2832736 CIM Client does not recognize HTTP extension headers

Changes in 1.3.9
==================
2807325 Change licensing from CPL to EPL

Changes in 1.3.8
==================
2414525 SLPConfig : parseList not returning populated list
2382765 HTTP header field Accept-Language does not include *
2372679 Add property to control synchronized SSL handshaking
2219646 Fix / clean code to remove compiler warnings

Changes in 1.3.7
==================
1984588 HttpClient not closed on cimclient close
2023050 DateTime object not accounting for microseconds correctly
1931266 M-POST not supported in java-client

Changes in 1.3.6
==================
1992321 1.3.6 packaging issues  
1954059 wrong path in CIMNameSpace URI
1954069 connection leak and exception in CIMClientXML
1960994 CIMSimpleDateTime.setDay() TODO and bad exception info
1960934 CIMDateTime(Calendar) does not respect DST
1901290 SLP error: "java.io.IOException" on Linux and IPv6
1913348 malformed service URL crashes SLP discovery
1931096 remove dependency of slpclient on cimclient classes
1931332 In HTTPClient need to get status before closing connection
1893499 no CIMExeption thrown for wrong namespace using SAXParser

Changes in 1.3.5
==================
1892041 Basic/digest authentication problem for Japanese users
1874440 sblimSLPclient PARSE ERROR
1835847 property for configuration file location
1832439 less strict parsing for IPv6 hostnames
1824094 CIMNameSpace(String pURI) doesn't handle namespace properly
1815752 TLS support
1816503 IPv6 support 
1804532 trace for both req/res should be traced in the same file
1804402 IPv6 ready SLP

Changes in 1.3.4
=================
1732645 Wrong reference building in METHODCALL request

Changes in 1.3.3
==================
1732645 Wrong reference building in METHODCALL request
1705776 Regression: parseBigInteger() doesn't like zero
1688270 Disable chunking because of trailer issues
1676343 Remove dependency from Xerces
1657901 Performance issues
1660568 Chunking broken on SUN JRE
1656285 IndicationHandler does not accept non-Integer message ID
1649779 Indication listener threads freeze

Changes in 1.3.2
==================
1649611 Interop issue: Quotes not escaped by client
1649595 No chunking requested
1647159 HttpClientPool runs out of HttpClients
1647148 HttpClient.resetSocket() doesn't set socket timeout
1646434 CIMClient close() invalidates all it's enumerations
1637546 CIMEnumerationImpl has faulty close function
1631407 VALUE.REFERENCE doesn't handle references without namespace
1627832 Incorrect retry behaviour on HTTP 401
1620526 Socket Leak in HTTPClient.getResponseCode()
1610046 Does not escape trailing spaces <KEYVALUE>
1604329 Fix OpenPegasus auth module
1516242 Support of OpenPegasus local authentication
1365086 Possible bug in createQualifier

Changes in 1.3.1
==================
1565091 ssl handshake exception
1574345 Client fails w/ NPE when processing chunked response
1573723 Selection of JSSE provider via properties file not feasible
1535793 Fix&Integrate CIM&SLP configuration classes
1558663 Support custom socket factories in client connections
1552457 NullPointer Exception while authenticating without PW
1547910 parseIMETHODCALL() CIMObjectPath parsing problem
1547908 parseKEYBINDING() creates incorrect reference type
1545915 Wrong parsing of IMETHODCALL request
1365082 Possible bugs in namespace creation
1535756 Make code warning free
1536711 NullPointerException causes client call to never return

Changes in 1.3
==================
 - #1528233: Interoperability with Engenio CIMOM broken
 - #1523854: deep always false with ecn
 - #1516246: Integrate SLP client code
 - #1516244: GCJ support
 - #1514405: getInstance() returns a keyless CIMInstance
 - #1498938: Multiple events in single cim-xml request are not handled
 - #1498927: Fill gaps in logging coverage
 - #1498130: Selection of xml parser on a per connection basis
 - #1493639: XML Parsing of PARAMETER.ARRAY wrong
 - #1488846: Bad format Locale information send to CIM Server
 - #1487705: sblimCIMClient throws NumberFormatException in UnsignedInt64
 - #1486379: CIM client retries twice when HTTP/1.1 401 is returned
 - #1381768: CIMClient.close() faulty on HTTPClientPool
 
Changes in 1.2.7-2
==================
 - #1483394: Indication listener threads don't close (fix integration)

Changes in 1.2.7
================
 - #1488924: Intermittent connection loss
 - #1483394: Indication listener threads don't close (broken fix)
 - #1483270: Using Several Cim Clients cause an indication problem
 - #1464860: No default value for VALUETYPE assumed
 - #1455939: CIMDataTime does not handle microsecond
 - #1438152: Wrong message ID in ExportResponseMessage
 - #1422316: Disable delayed acknowledgement

Changes in 1.2.6
================
 - #1377143: Mechanism to enable retries during transmission of request
 - #1381764: SBLIM Client getKey case sensitive
 - #1369307: CIM Client doesn't support embedded objects
 - #1375459: Upgrade for JPackage
 - #1365404: Log file gets created with logging disabled
 - #1362792: setUserPassword(char[]) does not clear old password
 - #1362783: PasswordCredential.getUserPassword expose char[] reference
 - #1362773: Possible NullPointerException in PassworCredential(char[])
 - #1353168: Possible NullPointerExcection in HttpClient.streamFinished()
 - #1353138: CIMObject element in HTTP header wrong formated
 - #1359805: UnsignedInt32.intValue() - Either Exception or method wrong
 - #1359684: Build: Separate SampleCode into own package


Changes in 1.2.5
================
 - #1339658: CIMEnumuration.finalize() closes used sockets

cim-client-java/src/main/java/org/sblim/slp/ServiceType.java
/**
 * ServiceType.java
 *
 * (C) Copyright IBM Corp. 2005, 2009
 *
 * THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE 
 * ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE 
 * CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.
 *
 * You can obtain a current copy of the Eclipse Public License from
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * @author: Roberto Pineiro, IBM, roberto.pineiro@us.ibm.com  
 * @author: Chung-hao Tan, IBM ,chungtan@us.ibm.com
 * 
 * Change History
 * Flag       Date        Prog         Description
 *------------------------------------------------------------------------------- 
 * 1516246    2006-07-22  lupusalex    Integrate SLP client code
 * 1804402    2007-09-28  ebak         IPv6 ready SLP
 * 2807325    2009-06-22  blaschke-oss Change licensing from CPL to EPL
 */


package org.sblim.slp;


import java.io.Serializable;
import java.util.NoSuchElementException;
import java.util.StringTokenizer;

/**
 * The ServiceType object models the SLP service type. It parses a string based
 * service type specifier into its various components, and contains property
 * accessors to return the components. URL schemes, protocol service types, and
 * abstract service types are all handled.
 */
public class ServiceType implements Serializable {

	private static final long serialVersionUID = -4850546870881037017L;

	private boolean iIsServiceURL = true;

	private String iPrincipleType = "";

	private String iAbstractType = "";

	private String iNamingAuthority = "";

	/**
	 * 
	 * Constructs a service type object from the service type specifier. Throws
	 * IllegalArgumentException if the type name is syntactically incorrect.
	 * 
	 * @param pType
	 *            The service type name as a String. If the service type is from
	 *            a service: URL, the "service:" prefix must be intact.
	 */
	public ServiceType(String pType) {
		parse(pType);
	}

	/**
	 * Returns true if the type name contains the "service:" prefix.
	 * 
	 * @return <code>true</code> if the type name contains the "service:"
	 *         prefix
	 */
	public boolean isServiceURL() {
		return iIsServiceURL;
	}

	/**
	 * Returns true if the type name is for an abstract type.
	 * 
	 * @return <code>true</code> if the type name is for an abstract type
	 */
	public boolean isAbstractType() {
		return iAbstractType.length() > 0;
	}

	/**
	 * Returns true if the naming authority is the default, i.e. is the empty
	 * string.
	 * 
	 * @return <code>true</code> if the naming authority is the default, i.e.
	 *         is the empty string
	 */
	public boolean isNADefault() {
		return iNamingAuthority.length() <= 0;
	}

	/**
	 * Returns the concrete type name in an abstract type, or the empty string
	 * if the service type is not abstract. For example, if the type name is
	 * "service:printing:ipp", the method returns "ipp". If the type name is
	 * "service:ftp", the method returns "".
	 * 
	 * @return <code>true</code> if the service type is not abstract
	 */
	public String getConcreteTypeName() {
		return iAbstractType;
	}

	/**
	 * Returns the abstract type name for an abstract type, the protocol name in
	 * a protocol type, or the URL scheme for a generic URL. For example, in the
	 * abstract type name "service:printing:ipp", the method returns "printing".
	 * In the protocol type name "service:ftp", the method returns "ftp".
	 * 
	 * @return The principle type name
	 */
	public String getPrincipleTypeName() {
		return iPrincipleType;
	}

	/**
	 * If the type is an abstract type, returns the fully formatted abstract
	 * type name including the "service:" and naming authority but without the
	 * concrete type name or intervening colon. If not an abstract type, returns
	 * the empty string. For example, in the abstract type name
	 * "service:printing:ipp", the method returns "service:printing".
	 * 
	 * @return The abstract type name
	 */
	public String getAbstractTypeName() {
		if (isAbstractType()) return "service:" + iPrincipleType
				+ (iNamingAuthority.length() <= 0 ? "" : "." + iNamingAuthority);
		return "";
	}

	/**
	 * Return the naming authority name, or the empty string if the naming
	 * authority is the default.
	 * 
	 * @return The naming authority
	 */
	public String getNamingAuthority() {
		return iNamingAuthority;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#equals(java.lang.Object)
	 * 
	 * Overrides Object.equals(). The two objects are equal if they are both
	 * ServiceType objects and the components of both are equal.
	 */
	public boolean equals(Object obj) {
		if (obj == this) return true;
		if (!(obj instanceof ServiceType)) return false;

		ServiceType servicetype = (ServiceType) obj;
		return iIsServiceURL == servicetype.iIsServiceURL
				&& iPrincipleType.equals(servicetype.iPrincipleType)
				&& iAbstractType.equals(servicetype.iAbstractType)
				&& iNamingAuthority.equals(servicetype.iNamingAuthority);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#toString()
	 * 
	 * Returns the fully formatted type name, including the "service:" if the
	 * type was originally from a service: URL.
	 */
	public String toString() {
		// TODO: clean up this
		return (iIsServiceURL ? "service:" : "") + iPrincipleType
				+ (iNamingAuthority.length() <= 0 ? "" : "." + iNamingAuthority)
				+ (iAbstractType.length() <= 0 ? "" : ":" + iAbstractType);
	}

	
	private int iHashCode = 0;
	
	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#hashCode()
	 * 
	 * Overrides Object.hashCode(). Hashes on the string value of the "service"
	 * prefix, naming authority, if any, abstract and concrete type names for
	 * abstract types, protocol type name for protocol types, and URL scheme for
	 * generic URLs.
	 */
	public int hashCode() {
		if (iHashCode == 0)
			iHashCode = toString().hashCode();
		return iHashCode;
	}

	private void parse(String pString) {
		StringTokenizer st = new StringTokenizer(pString, ":.", true);
		while (true) {
			try {

				String token = st.nextToken();
				if (token.equals(":") || token.equals(".")) continue;

				if (!token.equalsIgnoreCase("service")) {
					iIsServiceURL = false;
					do {
						iPrincipleType = iPrincipleType + token.toLowerCase();
						if (!st.hasMoreTokens()) break;
						token = st.nextToken();
					} while (true);

					validateTypeComponent(iPrincipleType);
					if (!st.hasMoreTokens()) return;
					continue;
				}
				token = st.nextToken();
				if (!token.equals(":")) continue;
				iPrincipleType = st.nextToken().toLowerCase();
				validateTypeComponent(iPrincipleType);

				if (!st.hasMoreTokens()) return;
				token = st.nextToken();
				if (token.equals(".")) {
					token = st.nextToken();
					validateTypeComponent(token);
					if (token.equalsIgnoreCase("iana")) continue;
					iNamingAuthority = token.toLowerCase();
					if (!st.hasMoreTokens()) return;
					token = st.nextToken();
				}
				if (token.equals(":")) {
					String abstractTypeToken = st.nextToken();
					validateTypeComponent(abstractTypeToken);
					iAbstractType = abstractTypeToken.toLowerCase();
					if (!st.hasMoreTokens()) return;
				}
			} catch (NoSuchElementException e) {
				throw new IllegalArgumentException(e.getMessage());
			}
		}
	}

	private static void validateTypeComponent(String str) {
		int length = str.length();

		for (int pos = 0; pos < length; pos++) {
			char ch = str.charAt(pos);
			if (!Character.isLetterOrDigit(ch) && ch != '+' && ch != '-') { throw new IllegalArgumentException(); }
		}
	}
}
